### DISCUSSION TITLE
[SetsMember] attribute

### DISCUSSION THREAD

**Discussion by TahirAhmadov** [dotnet/csharplang#6637] [Language Ideas]
As discussed in https://github.com/dotnet/csharplang/discussions/6405, https://github.com/dotnet/csharplang/discussions/6536, and https://github.com/dotnet/csharplang/discussions/6567, there seems to be a demand for a form of `[SetsRequiredMembers]` attribute which specifies which fields are set - in the context of [required properties](https://github.com/dotnet/csharplang/issues/3630). This got me thinking about something else - `init` methods - methods which can assign `readonly` fields and therefore can be used to "extract" code from ctors. 
```cs
class Foo
{
  public Foo()
  {
    this.Foo_Field1();
    this.Foo_Field2(); // etc. - complicated ctor is split into a few methods
    this._finalField = this._field1.Substring(this._field2.Length); // problem:
    // with NRT, compiler doesn't know _field1 and _field2 have been set
  }
  [SetsMember(nameof(_field1))] // this solves the problem above
  init void Foo_Field1() // future init method syntax
  {
    // complicated logic
    this._field1 = ...;
  }
  [SetsMember(nameof(_field2))] // this solves the problem above
  init void Foo_Field2() // future init method syntax
  {
    // complicated logic
    this._field2 = ...;
  }
  string _field1, _field2, _finalField;
}
```
Basically, I think we should add such an attribute, and make use of it for both the `required` properties ctor feature, as well as for the future `init` method feature. I think it's an elegant solution which solves multiple challenges - probably something other than what I described above, as well.

**Comment by TahirAhmadov** [comment]
Another option is to make it a keyword-based language feature:
```cs
class Foo
{
  public Foo()
  {
    this.Foo_Field1();
    this.Foo_Field2(); // etc. - complicated ctor is split into a few methods
    this._finalField = this._field1.Substring(this._field2.Length); // problem:
    // with NRT, compiler doesn't know _field1 and _field2 have been set
  }
  init void Foo_Field1() // future init method syntax
    set: _field1 // this solves the problem above
  {
    // complicated logic
    this._field1 = ...;
  }
  init void Foo_Field2() // future init method syntax
    set: _field2 // this solves the problem above
  {
    // complicated logic
    this._field2 = ...;
  }
  string _field1, _field2, _finalField;
}
```
Here, we repurpose the `set` keyword which comes after `where` clause (if applicable) and lists the members that are set by this method.

**Reply by 333fred** [reply]
This effectively reinvents the syntax we had thought about in earlier versions of the required members proposal. See the initial version of the proposal: https://github.com/dotnet/csharplang/blob/8a5955168e85fec63b404189cbbe8ca79c4fb34c/proposals/required-members.md#init-clauses.

**Reply by TahirAhmadov** [reply]
Yes, it kind of does; I haven't seen that before.
WRT the choice of keyword, I think using `init` would be confusing, if/when we get `init` methods, it would become weird:
```cs
init void Foo_Field1() : init(_field1)
{
}
```
In the same method definition, we have two meanings of the same keyword. 
Also, the word `init` kind of limits it to the context of "initialization", whereas it could be a valid concept in other contexts, as well - admittedly, none that I can think of right now.
However, I'm not stuck on the keyword choice in this matter.
The `init` clause for `required` properties only applies to ctors, whereas I envision that my `set` clause applies to both ctors and methods; and that allows us to "propagate" the setting of the members:
```cs
init void Foo_Field1AndField2() : set(_field1, _field2)
{
  this.Foo_Field1();
  this.Foo_Field2();
}
```

**Reply by 333fred** [reply]
We did not envision that the `init` clause would be limited to constructors, but would be potentially usable in `init` methods if we ever got them. This is part of why we scrapped the concept, actually: it's a complex metalanguage that, we hope, will have vanishingly small need to exist.

**Reply by TahirAhmadov** [reply]
> We did not envision that the init clause would be limited to constructors,

I got that impression because the `public C() : init(Prop!) => Initialize(1);` snippet shows a suppression syntax, which implies `Initialize` method couldn't get `init` clause, necessitating the suppression logic.

> will have vanishingly small need to exist.

I'm confused by this expectation. We already have the `required` properties, and hopefully `init` methods at some point. If anything, this increases the need for this feature, not vanishes.

**Reply by 333fred** [reply]
> If anything, this increases the need for this feature, not vanishes.

Our theory is that most people using `init` methods and constructor parameters aren't going to be using required members in the first place. I'm not particularly convinced that there will be a large number of users with required members that also want to have multiple constructors that set subsets of their members, much less a large number of users with enough of these constructors that they want to factor them into helper methods.

**Reply by TahirAhmadov** [reply]
WRT `required` properties, the benefit isn't just the propagation, but also the certainty of having a concrete list of members that are set in each particular ctor, as opposed to a blanket `[SetsRequiredMembers]` suppression (as was discussed at length elsewhere).
However, even outside the `required` properties, when it comes to any other fields and properties, it's not very often but common enough to have somewhat lengthy initialization for some of these members, and that's where the propagation aspect really shines. And this augments NRT analysis nicely.
I found [this comment](https://github.com/dotnet/csharplang/discussions/4704#discussioncomment-2061421) where you said it's better to use `[MemberNotNull]` for NRT analysis and while that solves the NRT problem, I think it's the less ideal solution. The "not null"-ness of the member is specified by its own declaration; we just need to communicate that it was set. The `[MemberNotNull]` has a different connotation - that a nullable member is now not null, due to some condition that was met - which is not what we're trying to express here. Think of it in terms of, you want to name your identifiers correctly; if a variable holding "surface area" is called "chocolateBunny" for some reason, the code will work, but we all agree that's poor naming; similarly here, `[MemberNotNull]` says something different than what we want to say - despite working technically.

**Reply by TonyValenti** [reply]
My biggest request with required members and constructors would be to allow constructors to "unrequire" specific members. This would be particularly useful with derived classes.
